cisco_aci/datadog_checks/cisco_aci/api.py

```diff
@@ -0,0 +1,243 @@
+
+class Sess:
+    def __init__(self, aci_url, session, apic_cookie, verify=None, timeout=None, log=None):
+        self.session = session
+        self.apic_cookie = apic_cookie
+        self.aci_url = aci_url
+        self.verify = verify
+        self.timeout = timeout
+        self.log = log
+
+    def send(self, req):
+        req.headers['Cookie'] = self.apic_cookie
+        return self.session.send(req, verify=self.verify, timeout=self.timeout)
+
+    def close(self):
+        self.session.close()
+
+    def make_request(self, path, raw_response=False, function_name=None):
+        url = "{0}{1}".format(self.aci_url, path)
+        req = Request('get', url)
+        prepped = req.prepare()
+        prepped.headers['Cookie'] = self.apic_cookie
+        response = self.session.send(prepped, verify=self.verify, timeout=self.timeout)
+        try:
+            response.raise_for_status()
+        except Exception as e:
+            self.log.warning("Error making request: {0}".format(e))
+            raise
+        path_save = path.replace('/', '_')
+        path_save = path_save.replace('?', '_')
+        path_save = path_save.replace('&', '_')
+        try:
+            if raw_response:
+                return response
+            else:
+                return response.json()
+        except Exception as e:
+            self.log.warning("Exception in json parsing, returning nothing: {0}".format(e))
+            raise
+
+
+class Api:
+    def __init__(self, aci_urls, username, password, verify=False, timeout=10, log=None):
+        self.aci_urls = aci_urls
+        self.username = username
+        self.password = password
+        self.timeout = timeout
+        self.verify = verify
+        self.sessions = []
+        if log:
+            self.log = log
+        else:
+            import logging
+            self.log = logging.getLogger('cisco_api')
+
+    def close(self):
+        for session in self.sessions:
+            session.close()
+
+    def login(self):
+        # ensure sessions are an empty array
+        self.sessions = []
+        for aci_url in self.aci_urls:
+            session = Session()
+            data = '<aaaUser name="{0}" pwd="{1}"/>\n'.format(self.username, self.password)
+            url = "{0}{1}".format(aci_url, '/api/aaaLogin.xml')
+            req = Request('post', url, data=data)
+            prepped_request = req.prepare()
+            response = session.send(prepped_request, verify=self.verify, timeout=self.timeout)
+            response.raise_for_status()
+            self.apic_cookie = 'APIC-Cookie={}'.format(response.cookies.get('APIC-cookie'))
+            sess = Sess(aci_url, session, self.apic_cookie,
+                        verify=self.verify,
+                        timeout=self.timeout,
+                        log=self.log)
+            self.sessions.append(sess)
+
+    def make_request(self, path, raw_response=False, function_name=None):
+        session = random.choice(self.sessions)
```

Comment:
what's the rationale behind this? can we add a note?

Reply:
I added one